src/device/ping-device.cpp

```diff
@@ -0,0 +1,94 @@
+#include <ping-device.h>
+
+#include <ping-message-common.h>
+#include <ping-message.h>
+#include <ping-time.h>
+
+bool PingDevice::initialize() {
+    return request(CommonId::PROTOCOL_VERSION)
+        && request(CommonId::DEVICE_INFORMATION);
+}
+
+ping_message* PingDevice::read() {
+    char b;
+    int result = _port.read(&b, 1);
+    if (result > 0) {
+        if (_parser.parseByte(static_cast<char>(b)) == PingParser::NEW_MESSAGE) {
+        return &_parser.rxMessage;
+        }
+    }
+    return nullptr;
+}
+
+ping_message* PingDevice::request(uint16_t id, int timeoutMs) {
+    common_general_request msg;
+    msg.set_requested_id(id);
+    writeMessage(msg);
+    ping_message* reply = waitMessage(id, timeoutMs);
+    if (reply && reply->message_id() == CommonId::NACK) {
+        common_nack* m = static_cast<common_nack*>(reply);
+        if (m->nacked_id() != CommonId::GENERAL_REQUEST) {
+        return nullptr;
+        }
+    }
+
+    return reply;
+}
+
```

Comment:
can the id be identified by an enum? a integer is too error prone for this. 
Why are you blocking in the waitMessage? this looks like it should be written with async in mind.
nitpick: code intentation

Reply:
a message id is by definition a uint16, so it is safe. We can not type the argument as enum because we do not have a single enum space for the messages.

enums may be submitted as arguments without issue and to avoid errors.

This is a migration of a synchronous implementation already in place. to move to async support we will have to do more development. The synchronous implementation serves us well for demonstration purposes.

Reply:
See copy elision